File: apps/mobile/lib/features/tasks/application/recurrence.dart
import 'dart:convert';

import 'package:timezone/timezone.dart' as tz;

import '../../../core/notifications/alert_plan.dart' show memberZone;

/// A repeating task's rule, as the row stores it.
///
/// The rule and its exceptions, never expanded rows: `dtstart`, an RRULE
/// string, the mode that decides what "next" means, and the occurrences the
/// member removed. Storing occurrences would make a moved one an edit to the
/// series instead of an override, and a series has no bottom.
class Recurrence {
  const Recurrence({
    required this.dtstart,
    required this.rrule,
    required this.mode,
    this.exdates = const [],
  });

  /// From the `recurrenceJson` column, or null when there is no rule or the
  /// JSON is not one.
  ///
  /// Null rather than a throw: an unreadable rule reaches here from a row a
  /// newer client wrote, and one bad rule must not take a list render down.
  static Recurrence? decode(String? encoded) {
    if (encoded == null || encoded.isEmpty) return null;
    try {
      final decoded = jsonDecode(encoded);
      if (decoded is! Map) return null;
      final dtstart = DateTime.tryParse('${decoded['dtstart']}');
      final rrule = decoded['rrule'];
      if (dtstart == null || rrule is! String || rrule.isEmpty) return null;
      return Recurrence(
        dtstart: dtstart.toUtc(),
        rrule: rrule,
        mode: decoded['mode'] == 'completion'
            ? RecurrenceMode.completion
            : RecurrenceMode.schedule,
        exdates: [
          for (final raw in (decoded['exdates'] as List? ?? const []))
            if (DateTime.tryParse('$raw') != null)
              DateTime.parse('$raw').toUtc(),
        ],
      );
    } catch (_) {
      return null;
    }
  }

  final DateTime dtstart;
  final String rrule;
  final RecurrenceMode mode;
  final List<DateTime> exdates;

  String encode() => jsonEncode({
    'dtstart': dtstart.toUtc().toIso8601String(),
    'rrule': rrule,
    'mode': mode.name,
    'exdates': [for (final d in exdates) d.toUtc().toIso8601String()],
  });

  /// `FREQ=WEEKLY;INTERVAL=2` → the pieces, upper-cased and keyed.
  Map<String, String> get parts => {
    for (final piece in rrule.split(';'))
      if (piece.contains('='))
        piece.split('=').first.trim().toUpperCase():
            piece.split('=').last.trim().toUpperCase(),
  };

  /// The next occurrence strictly after a completion, or null when there is
  /// none — the series ended, or this build cannot read the rule.
  ///
  /// The two modes are the member's own words, and the difference is the whole
  /// reason the field exists:
  ///
  /// * **schedule** — "water the plants every Tuesday". Tuesday comes whether or
  ///   not last Tuesday's watering happened, so the step is measured from the
  ///   *scheduled* occurrence.
  /// * **completion** — "every two weeks from when I actually do it". The step
  ///   is applied to the day the work was done, keeping the series' own time of
  ///   day: a task ticked off at 23:50 must not drift into the night.
  ///
  /// The arithmetic is done on calendar fields in the member's own zone rather
  /// than by adding a [Duration]. A day is not always 24 hours, and "every
  /// Tuesday at 08:00" across a daylight-saving boundary lands at 07:00 or
  /// 09:00 if it is — which is exactly the shift the member is least likely to
  /// forgive.
  ///
  /// ponytail: FREQ and INTERVAL and UNTIL, which is precisely what the repeat
  /// picker can produce. A rule carrying BYDAY, BYMONTHDAY or COUNT — one the
  /// web app or the chat extractor made — answers null, and the task then
  /// simply completes without advancing. Upgrade to a real RRULE package on the
  /// day one of those rules can be created on the phone; until then the
  /// dependency would parse grammar nothing here can write.
  DateTime? nextAfterCompleting({
    required DateTime completedAt,
    required DateTime? dueAt,
    String? timezone,
  }) {
    final fields = parts;
    if (fields.containsKey('BYDAY') ||
        fields.containsKey('BYMONTHDAY') ||
        fields.containsKey('COUNT')) {
      return null;
    }
    final interval = int.tryParse(fields['INTERVAL'] ?? '1') ?? 1;
    if (interval < 1) return null;
    final freq = fields['FREQ'];
    if (freq == null) return null;

    final zone = memberZone(timezone);
    final from = tz.TZDateTime.from(
      mode == RecurrenceMode.completion
          ? completedAt
          : (dueAt ?? dtstart),
      zone,
    );
    // In completion mode the step lands on the day the work was done, at the
    // series' own hour — not at the hour it happened to be ticked off.
    final timeOf = tz.TZDateTime.from(dueAt ?? dtstart, zone);
    final until = _until(fields['UNTIL']);

    // Walked rather than taken in one step, because an exdate is an occurrence
    // that was removed and the one after it is what the member expects next.
    // Two hundred steps is four years of weekly; beyond that the rule is not
    // one this build should be guessing at.
    for (var step = 1; step <= 200; step++) {
      final candidate = _advance(from, timeOf, freq, interval * step, zone);
      if (candidate == null) return null;
      final instant = DateTime.fromMillisecondsSinceEpoch(
        candidate.millisecondsSinceEpoch,
        isUtc: true,
      );
      if (!instant.isAfter(completedAt.toUtc())) continue;
      if (until != null && instant.isAfter(until)) return null;
      if (exdates.any((e) => e.isAtSameMomentAs(instant))) continue;
      return instant;
    }
    return null;
  }

  static DateTime? _until(String? raw) {
    if (raw == null || raw.isEmpty) return null;
    // `20261231T235959Z`, the RRULE spelling. Punctuated back into something
    // `DateTime.parse` accepts rather than parsed by hand.
    final match = RegExp(
      r'^(\d{4})(\d{2})(\d{2})(?:T(\d{2})(\d{2})(\d{2})Z?)?$',
    ).firstMatch(raw);
    if (match == null) return DateTime.tryParse(raw)?.toUtc();
    final g = match.groups([1, 2, 3, 4, 5, 6]);
    return DateTime.utc(
      int.parse(g[0]!),
      int.parse(g[1]!),
      int.parse(g[2]!),
      int.parse(g[3] ?? '0'),
      int.parse(g[4] ?? '0'),
      int.parse(g[5] ?? '0'),
    );
  }

  /// [count] steps of [freq] from [from], at [timeOf]'s hour.
  ///
  /// Built from calendar fields — `tz.TZDateTime(zone, y, m, d, h, min)` — so
  /// the zone library resolves the wall clock, including a day that is 23 or 25
  /// hours long and a time that a daylight-saving gap skips.
  static tz.TZDateTime? _advance(
    tz.TZDateTime from,
    tz.TZDateTime timeOf,
    String freq,
    int count,
    tz.Location zone,
  ) => switch (freq) {
    'DAILY' => tz.TZDateTime(
      zone,
      from.year,
      from.month,
      from.day + count,
      timeOf.hour,
      timeOf.minute,
    ),
    'WEEKLY' => tz.TZDateTime(
      zone,
      from.year,
      from.month,
      from.day + count * 7,
      timeOf.hour,
      timeOf.minute,
    ),
    'MONTHLY' => tz.TZDateTime(
      zone,
      from.year,
      from.month + count,
      from.day,
      timeOf.hour,
      timeOf.minute,
    ),
    'YEARLY' => tz.TZDateTime(
      zone,
      from.year + count,
      from.month,
      from.day,
      timeOf.hour,
      timeOf.minute,
    ),
    _ => null,
  };
}

/// What "next" is measured from. See [Recurrence.nextAfterCompleting].
enum RecurrenceMode { schedule, completion }

/// The repeats the editor offers, as the RRULE each one means.
///
/// A fixed set rather than a rule builder: these four cover what a task list is
/// for, and every one of them is a rule [Recurrence.nextAfterCompleting] can
/// advance without a parser. Offering a grammar the phone cannot walk would be
/// offering a repeat that silently stops repeating.
const Map<String, String> kRepeatChoices = {
  'Every day': 'FREQ=DAILY;INTERVAL=1',
  'Every week': 'FREQ=WEEKLY;INTERVAL=1',
  'Every 2 weeks': 'FREQ=WEEKLY;INTERVAL=2',
  'Every month': 'FREQ=MONTHLY;INTERVAL=1',
  'Every year': 'FREQ=YEARLY;INTERVAL=1',
};

File: apps/backend/src/contexts/planning/domain/recurrence.ts
import { RRule, rrulestr } from 'rrule';
import {
  localDate,
  localHhMm,
  wallClockToUtc,
} from '../../../shared/time/time.js';

export type RecurrenceMode = 'schedule' | 'completion';

/**
 * A repeating task's rule, as stored: a start, an RRULE string, the mode that
 * decides what "next" means, and the occurrences the member removed.
 *
 * Never expanded into rows. A recurring task is one document; the occurrences
 * are computed for the window somebody asked about, and a moved occurrence is
 * an override rather than an edit to the series.
 */
export interface RecurrenceRule {
  dtstart: Date;
  rrule: string;
  mode: RecurrenceMode;
  exdates: Date[];
}

/**
 * The two modes, in the member's words:
 *
 *   - **schedule** — "water the plants every Tuesday". Tuesday comes whether or
 *     not last Tuesday's watering happened. The next occurrence is measured
 *     from the *scheduled* one.
 *   - **completion** — "change the bedsheets every two weeks". Two weeks from
 *     when you actually did it, not from when you were supposed to. The next
 *     occurrence is measured from the completion.
 *
 * They differ only when a task is completed late or early, which is most of the
 * time, and getting it wrong is the difference between a chore list that drifts
 * with your life and one that piles up accusingly.
 */

/**
 * Every recurrence computation happens in the member's own zone.
 *
 * This is not a nicety. `rrule` has no notion of a time zone: it treats a Date
 * as a bag of fields and steps them forward, so handing it a real instant makes
 * every occurrence a fixed *UTC* offset from the first one. A weekly task set
 * for 18:00 in Cairo would then fire at 18:00 through the winter and 17:00
 * after the clocks moved — the same class of bug that once shifted every
 * extracted reminder by three hours, arriving by a different route.
 *
 * So the rule is evaluated on *floating* dates: a Date whose UTC fields hold
 * the member's wall clock. `rrule` steps those, and each result is converted
 * back to a real instant through `shared/time`, which knows what the member's
 * clock read that day. "Every Tuesday at 18:00" then means 18:00 on the
 * member's clock every Tuesday, across a daylight-saving change, which is what
 * the member asked for.
 *
 * Minute granularity: seconds are dropped on the way through. Nothing in this
 * product schedules to the second, and carrying a field no feature reads would
 * only be a place for the two representations to disagree.
 */
function toFloating(instant: Date, timezone: string): Date {
  const date = localDate(instant, timezone); // YYYY-MM-DD in the member's zone
  const time = localHhMm(instant, timezone); // HH:mm on the member's clock
  const [year, month, day] = date.split('-').map(Number);
  const [hour, minute] = time.split(':').map(Number);
  return new Date(
    Date.UTC(year ?? 1970, (month ?? 1) - 1, day ?? 1, hour ?? 0, minute ?? 0),
  );
}

/**
 * The inverse. A floating Date's UTC fields are read as a wall clock and
 * resolved against the zone, so an occurrence landing in a daylight-saving gap
 * comes back as the first valid instant after it rather than as an hour that
 * never happened.
 */
function fromFloating(floating: Date, timezone: string): Date | null {
  const wallClock =
    `${floating.getUTCFullYear()}-` +
    `${String(floating.getUTCMonth() + 1).padStart(2, '0')}-` +
    `${String(floating.getUTCDate()).padStart(2, '0')}T` +
    `${String(floating.getUTCHours()).padStart(2, '0')}:` +
    `${String(floating.getUTCMinutes()).padStart(2, '0')}`;
  return wallClockToUtc(wallClock, timezone);
}

/** Same minute on the member's clock, which is how an exdate is matched. */
function sameMinute(a: Date, b: Date): boolean {
  return Math.floor(a.getTime() / 60_000) === Math.floor(b.getTime() / 60_000);
}

/**
 * How far apart two occurrences of this rule are, as a calendar step rather
 * than a duration.
 *
 * A duration would be wrong twice over: adding 24 hours to an instant crosses a
 * daylight-saving boundary an hour off, and "every month" has no fixed length
 * at all. So the step is expressed in the unit the rule is written in and
 * applied to the member's local calendar date, where "a month later" means the
 * same day number in the next month and "a day later" means tomorrow whatever
 * happened to the clocks.
 */
interface CalendarStep {
  unit: 'day' | 'week' | 'month' | 'year';
  count: number;
}

function stepOf(rule: RRule): CalendarStep {
  const count = rule.options.interval || 1;
  switch (rule.options.freq) {
    case RRule.YEARLY:
      return { unit: 'year', count };
    case RRule.MONTHLY:
      return { unit: 'month', count };
    case RRule.WEEKLY:
      return { unit: 'week', count };
    default:
      // DAILY, and anything finer. Sub-daily recurrence is not a thing this
      // product offers — the editor only writes daily, weekly, monthly and
      // yearly rules — so an HOURLY rule arriving from somewhere is treated as
      // daily rather than accepted as a rule nothing else in the system can
      // render or explain to the member.
      return { unit: 'day', count };
  }
}

/**
 * Advances a local calendar date, then re-imposes a wall-clock time.
 *
 * The clamp on month and year steps is the one every calendar has to make: the
 * 31st plus one month has no answer in February, and the choice here is the
 * last day of the shorter month rather than spilling into March. A monthly task
 * set for the 31st therefore lands on the 28th, 29th or 30th and then returns
 * to the 31st, because the step is always measured from the rule's own start
 * rather than from wherever the last clamp landed.
 */
function addStep(
  fromLocalDate: string,
  step: CalendarStep,
  occurrence: number,
): string {
  const [year, month, day] = fromLocalDate.split('-').map(Number);
  const y = year ?? 1970;
  const m = (month ?? 1) - 1;
  const d = day ?? 1;
  const amount = step.count * occurrence;

  if (step.unit === 'day' || step.unit === 'week') {
    const days = step.unit === 'week' ? amount * 7 : amount;
    const moved = new Date(Date.UTC(y, m, d + days));
    return isoDate(moved);
  }

  const months = step.unit === 'year' ? amount * 12 : amount;
  const targetMonth = m + months;
  const targetYear = y + Math.floor(targetMonth / 12);
  const normalisedMonth = ((targetMonth % 12) + 12) % 12;
  // Day 0 of the following month is the last day of this one.
  const lastDay = new Date(
    Date.UTC(targetYear, normalisedMonth + 1, 0),
  ).getUTCDate();
  return isoDate(
    new Date(Date.UTC(targetYear, normalisedMonth, Math.min(d, lastDay))),
  );
}

function isoDate(at: Date): string {
  return `${at.getUTCFullYear()}-${String(at.getUTCMonth() + 1).padStart(2, '0')}-${String(
    at.getUTCDate(),
  ).padStart(2, '0')}`;
}

/**
 * The rule, parsed once and asked questions.
 *
 * Constructed through `parse`, which returns null for a rule string the library
 * refuses. A task carrying an unparseable rule is a task that cannot be
 * completed — the advance would throw on the one operation the member most
 * wants to succeed — so the aggregate treats a refused rule as "no recurrence"
 * and says so at the boundary instead.
 */
export class Recurrence {
  private constructor(
    private readonly rule: RRule,
    readonly spec: RecurrenceRule,
  ) {}

  /**
   * `null` rather than a throw, because the caller is a domain method with a
   * better answer available: refuse the write at the edge, where the member can
   * be told, rather than half-way through completing a task.
   */
  static parse(spec: RecurrenceRule, timezone: string): Recurrence | null {
    try {
      const parsed = rrulestr(
        spec.rrule.startsWith('RRULE:') || spec.rrule.startsWith('DTSTART')
          ? spec.rrule
          : `RRULE:${spec.rrule}`,
        { dtstart: toFloating(spec.dtstart, timezone) },
      );
      // `rrulestr` also parses RRULESET, which carries its own exdates and its
      // own semantics for them. This context stores exceptions in `exdates` and
      // applies them here, so a set arriving from outside would have two
      // exception lists disagreeing about the same series.
      if (!(parsed instanceof RRule)) return null;
      return new Recurrence(parsed, spec);
    } catch {
      return null;
    }
  }

  get mode(): RecurrenceMode {
    return this.spec.mode;
  }

  /**
   * The next occurrence strictly after `after`, or null when the series has
   * run out — a rule with `COUNT` or `UNTIL` ends, and a task whose series has
   * ended simply completes for the last time.
   *
   * In **schedule** mode this is the rule's own next occurrence: the calendar
   * decides, and completing late does not move the series.
   *
   * In **completion** mode the rule supplies only the step; the step is applied
   * to the day the work was actually done, keeping the series' own time of day.
   * "Every two weeks from when I do it" is then two weeks from the doing, at
   * the hour the member originally chose — not at whatever hour they happened
   * to tick it off, which would let a task done at 23:50 drift into the night.
   */
  next(after: Date, timezone: string): Date | null {
    if (this.spec.mode === 'completion')
      return this.nextFromCompletion(after, timezone);

    const floatingAfter = toFloating(after, timezone);
    // Walk forward rather than take the first hit, because an exdate is an
    // occurrence the member removed and the one after it is the answer. The
    // bound is a guard against a rule whose every remaining occurrence has been
    // excluded: 200 is far past any real exdate list and stops a pathological
    // rule from spinning.
    let cursor = floatingAfter;
    for (let attempt = 0; attempt < 200; attempt += 1) {
      const occurrence = this.rule.after(cursor, false);
      if (!occurrence) return null;

      const instant = fromFloating(occurrence, timezone);
      if (!instant) return null;
      if (!this.isExcluded(instant)) return instant;
      cursor = occurrence;
    }
    return null;
  }

  /**
   * Mode `completion`: the step applied to the completion's own local day, with
   * the series' time of day re-imposed.
   *
   * The time of day comes from `dtstart` and not from the completion, which is
   * the whole point — see `next`.
   */
  private nextFromCompletion(completedAt: Date, timezone: string): Date | null {
    const step = stepOf(this.rule);
    const timeOfDay = localHhMm(this.spec.dtstart, timezone);

    // Occurrence 1 is one step out. If the member removed that day, take the
    // next step, and so on: the exception list means "not this day", and it
    // means it whichever mode moved us there.
    for (let occurrence = 1; occurrence <= 200; occurrence += 1) {
      const day = addStep(localDate(completedAt, timezone), step, occurrence);
      const instant = wallClockToUtc(`${day}T${timeOfDay}`, timezone);
      if (!instant) return null;
      if (!this.isExcluded(instant)) return instant;
    }
    return null;
  }

  /**
   * Every occurrence inside a window, for a caller rendering a calendar.
   *
   * `completion`-mode series have no calendar of their own — where the next one
   * lands depends on when the current one is done, which has not happened yet —
   * so such a series reports only the occurrence it is currently sitting on.
   * Projecting a speculative chain onto a month view would show the member
   * appointments they never made.
   */
  between(from: Date, to: Date, timezone: string): Date[] {
    if (this.spec.mode === 'completion') return [];

    return this.rule
      .between(toFloating(from, timezone), toFloating(to, timezone), true)
      .map((occurrence) => fromFloating(occurrence, timezone))
      .filter(
        (instant): instant is Date =>
          instant !== null && !this.isExcluded(instant),
      );
  }

  /**
   * The rule with one occurrence removed. Returns the new spec rather than
   * mutating, because the aggregate is what decides to keep it — skipping an
   * occurrence is an event, not a side effect of asking a question.
   */
  skip(occurrence: Date): RecurrenceRule {
    if (this.isExcluded(occurrence)) return this.spec;
    return { ...this.spec, exdates: [...this.spec.exdates, occurrence] };
  }

  isExcluded(occurrence: Date): boolean {
    return this.spec.exdates.some((exdate) => sameMinute(exdate, occurrence));
  }

  /**
   * "every 2 weeks on Tuesday", for a confirmation line.
   *
   * English only, and deliberately not localised here: `rrule`'s own
   * translation needs a gettext table per language, and the client already
   * holds the strings for its locale. The server sends the rule; the phone
   * renders it in Arabic when that is what the member reads.
   */
  humanText(): string {
    try {
      return this.rule.toText();
    } catch {
      return this.spec.rrule;
    }
  }
}
